### PR TITLE
Fix WebSocket negotiation

### PR DESCRIPTION
**Pull request by RX14**
I have also added some specs (with HTTP and HTTPS) for this.

### CODE REVIEW

File: spec/std/http/web_socket_spec.cr [hunk 1]
```diff
@@ -293,6 +293,84 @@ describe HTTP::WebSocket do
...
+
+      http_server = http_ref = HTTP::Server.new(0, [ws_handler])
+      http_server.bind
+      port_chan.send(http_server.port)
+      http_server.listen
+    end
+
+    listen_port = port_chan.receive
+
+    ws2 = HTTP::WebSocket.new("ws://127.0.0.1:#{listen_port}")
```
Comment (RX14) [inline spec/std/http/web_socket_spec.cr]: I consider it a bug that I have to use `ws2` to differentiate this variable from `ws` above, as *that* ws is introduced as a block parameter, which should introduce an entirely new variable.

Comment (bcardiff) [inline spec/std/http/web_socket_spec.cr]: I agree, it should not be required to use another variable. It seems that the `on_message` in the top is been triggered with the websocket in the bottom. between the two closures that covers `ws` and the `spawn` there might be something going on.

File: spec/std/http/web_socket_spec.cr [hunk 2]
```diff
@@ -293,6 +293,84 @@ describe HTTP::WebSocket do
...
+      ws_handler = HTTP::WebSocketHandler.new do |ws, ctx|
+        ctx.request.path.should eq("/")
+        ws.on_message do |str|
+          ws.send("pong #{str}")
+        end
+
+        ws.on_close do
+          http_ref.not_nil!.close
+        end
+        ws.run
```
Comment (bcardiff) [inline spec/std/http/web_socket_spec.cr]: on the server side there should be no `ws.run`

Comment (RX14) [inline spec/std/http/web_socket_spec.cr]: Ah thanks, I forgot it did it for me.

Comment (RX14) [inline spec/std/http/web_socket_spec.cr]: Fixed.

**Comment by RX14** [comment]
I think that specs that run a `HTTP::Server` should be fine as long as we let the OS choose a port. There *appear* to be specs that run `TCPServer`, so I don't see that there would be any issue.

**Comment by bcardiff** [comment]
Thanks @RX14 . You were right regarding TCPServer and letting OS choose a port.

**Pull request merged**